cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -164,31 +372,303 @@ theorem mapOnValues_contains {α β γ} [LT α] [DecidableLT α] [DecidableEq α
-    omega
-  rename_i v_lt_kv m1_lt_m
-  let a := sizeOf v
-  let c := sizeOf m.1
-  let d := sizeOf m
-  have v_lt_m1 : a < c := by apply Nat.lt_trans v_lt_kv h
-  have v_lt_m : a < d := by apply Nat.lt_trans v_lt_m1 m1_lt_m
-  have ha : a = sizeOf v := by simp
-  have hd : d = sizeOf m := by simp
-  rw [ha, hd] at v_lt_m
-  exact v_lt_m
+  simp only [values, List.mem_map, findOrErr_ok_iff_find?_some]
+  constructor
+  case mp =>
+    intro h₁
+    replace ⟨⟨k, v'⟩, ⟨h₁, h₂⟩⟩ := h₁
+    simp only at h₂
+    subst v'
+    exists k
+    simp [h₁, ← in_list_some_find? wf]
+  case mpr =>
+    intro h₁
+    replace ⟨k, h₁⟩ := h₁
+    exists (k, v)
+    simp [h₁, in_list_some_find? wf, and_true]
+
+theorem in_kvs_in_mapOnValues [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} {v : β} :
+  (k, v) ∈ m.kvs → (k, f v) ∈ (m.mapOnValues f).kvs
+:= by
+  unfold mapOnValues
+  intro h₁
+  simp only [kvs, List.mem_map, Prod.mk.injEq]
+  exists (k, v)
+
+/-! ### mapMOnValues -/
+
+/--
+  This is not stated in terms of `Map.keys` because `Map.keys` produces a `Set`,
+  and we want the even stronger property that it not only preserves the key-set,
+  but also the key-order. (We'll use this to prove `mapMOnValues_wf`.)
+-/
+theorem mapMOnValues_preserves_keys [LT α] [DecidableLT α] [StrictLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  m₁.kvs.map Prod.fst = m₂.kvs.map Prod.fst
+:= by
+  intro h₁
+  simp only [mapMOnValues, Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some,
+    Option.some.injEq] at h₁
+  replace ⟨xs, h₁, h₂⟩ := h₁
+  subst h₂
+  cases h₂ : m₁.kvs <;> simp only [h₂, List.mapM_nil, List.mapM_cons, Option.pure_def,
+    Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq] at h₁
+  <;> unfold kvs at *
+  case nil =>
+    subst h₁
+    simp [h₂]
+  case cons kv tl =>
+    have (k, v) := kv ; clear kv
+    replace ⟨(k', y), ⟨y', h₁, h₃⟩, ⟨tl', h₄, h₅⟩⟩ := h₁
+    subst h₅
+    simp only [Prod.mk.injEq, List.map_cons, List.cons.injEq] at *
+    replace ⟨h₃, h₃'⟩ := h₃
+    subst k' y'
+    have ih := mapMOnValues_preserves_keys (m₁ := mk tl) (m₂ := mk tl') (f := f)
+    simp only [mapMOnValues, kvs, Option.pure_def, Option.bind_eq_bind,
+      Option.bind_eq_some, Option.some.injEq, mk.injEq, exists_eq_right] at ih
+    specialize ih h₄
+    simp [ih, h₂]
+
+theorem mapMOnValues_wf [LT α] [DecidableLT α] [StrictLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.WellFormed →
+  (m₁.mapMOnValues f = some m₂) →
+  m₂.WellFormed
+:= by
+  simp only [wf_iff_sorted, toList]
+  intro wf h₁
+  have h₂ := mapMOnValues_preserves_keys h₁
+  exact (List.map_eq_implies_sortedBy h₂).mp wf
+
+/--
+  Alternate proof of `mapMOnValues_wf`, that relies on
+  `List.mapM_some_eq_filterMap` instead of `mapMOnValues_preserves_keys`. Which do
+  we prefer?
+-/
+theorem mapMOnValues_wf_alt_proof [LT α] [DecidableLT α] [StrictLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
```

Comment:
Thread for discussion of the two proofs of `mapMOnValues_wf`

Reply:
I don't have strong feelings on this one.  The `mapMOnValues_wf` proof is shorter than the `alt` version because it pushes the complexity into `mapMOnValues_preserves_keys`.  If this is the only use case for `mapMOnValues_preserves_keys`, I'd be tempted to go for the `alt` proof and delete the special-purpose lemma.

Reply:
This is the only use case for `mapMOnValues_preserves_keys`, but my hesitation is that since we've gone through the trouble to prove it now, we might want to keep it in case it is useful in the future.  Or maybe it's just super unlikely to be useful in the future?

Reply:
Not sure.  I haven't needed it in the large internal coebase :). But I don't see harm in keeping it.